app/models/bitbucket_integration.rb

```diff
@@ -0,0 +1,379 @@
+  end
+
+  def tag_url(tag_name)
+    "https://bitbucket.org/#{code_repository_name}/src/#{tag_name}"
+  end
+
+  def compare_url(to_branch, from_branch)
+    "https://bitbucket.org/#{code_repository_name}/compare/#{to_branch}..#{from_branch}"
+  end
+
+  def pull_requests_url(branch_name, open: false)
+    state = open ? "OPEN" : "ALL"
+    q = URI.encode_www_form("state" => state, "at" => branch_name)
+    "https://bitbucket.org/#{code_repository_name}/pull-requests/?#{q}"
+  end
+
+  def tag_exists?(tag_name)
+    with_api_retries { installation.get_tag(code_repository_name, tag_name) }.present?
+  rescue Installations::Error => ex
+    raise ex unless ex.reason == :not_found
+    false
+  end
+
+  def create_release!(tag_name, branch_name) = create_tag!(tag_name, branch_name)
+
+  def create_tag!(tag_name, sha)
+    with_api_retries { installation.create_tag!(code_repository_name, tag_name, sha) }
+  end
+
+  def create_pr!(to_branch_ref, from_branch_ref, title, description)
+    with_api_retries do
+      installation
+        .create_pr!(code_repository_name, to_branch_ref, from_branch_ref, title, description, PR_TRANSFORMATIONS)
+        .merge_if_present(source: :bitbucket)
+    end
+  end
+
+  def find_pr(to_branch_ref, from_branch_ref)
+    with_api_retries do
+      installation
+        .find_pr(code_repository_name, to_branch_ref, from_branch_ref, PR_TRANSFORMATIONS)
+        .merge_if_present(source: :bitbucket)
+    end
+  end
+
+  def get_pr(pr_number)
+    with_api_retries { installation.get_pr(code_repository_name, pr_number, PR_TRANSFORMATIONS).merge_if_present(source: :bitbucket) }
+  end
+
+  def merge_pr!(pr_number)
+    with_api_retries { installation.merge_pr!(code_repository_name, pr_number) }
+  end
+
+  def create_patch_pr!(_to_branch, _patch_branch, _commit_hash, _pr_title_prefix)
+    raise NotImplementedError
+  end
+
+  def enable_auto_merge!(_pr_number)
+    raise NotImplementedError
+  end
+
+  # CI/CD
+
+  WORKFLOW_TRANSFORMATIONS = {
+    id: :id,
+    name: :name
+  }
+
+  WORKFLOW_RUN_TRANSFORMATIONS = {
+    ci_ref: :uuid,
+    number: :build_number
+  }
+
+  ARTIFACTS_TRANSFORMATIONS = {
+    id: :name,
+    name: :name,
+    size_in_bytes: :size,
+    archive_download_url: [:links, :self, :href],
+    generated_at: :created_on
+  }
+
+  def workflows
+    return [] unless integration.ci_cd?
+    with_api_retries { installation.list_pipeline_selectors(code_repository_name, WORKFLOW_TRANSFORMATIONS) }
+  end
+
+  def workflow_retriable? = false
+
+  def trigger_workflow_run!(ci_cd_channel, _branch_name, inputs, commit_hash = nil)
+    with_api_retries do
+      res = installation.trigger_pipeline!(code_repository_name, ci_cd_channel, inputs, commit_hash, WORKFLOW_RUN_TRANSFORMATIONS)
+      res.merge(ci_link: "https://bitbucket.org/#{code_repository_name}/pipelines/results/#{res[:number]}")
+    end
+  end
+
+  def retry_workflow_run!(_ci_ref)
+    raise Integrations::UnsupportedAction
```

Comment:
Implement this

Reply:
Bitbucket doesn't support retrying the existing pipeline, instead, It creates a new pipeline. So, this method cannot be implemented.

We handle this by letting `WorkflowRun` know by making `BitbucketIntegration` as not `workflow_retriable`.